File: android/core/network/src/main/kotlin/com/devora/mencare/core/network/dto/AdminDto.kt
package com.devora.mencare.core.network.dto

import kotlinx.serialization.Serializable
import kotlinx.serialization.json.JsonElement

/** Area titolare: cruscotto, regole delle notifiche e campagne push (`/v1/admin/…`). */

@Serializable
data class OperatorOccupancyDto(val operatorId: String, val operatorName: String, val percent: Int)

@Serializable
data class UpcomingDayDto(
    val date: String,
    val occupancyPercent: Int = 0,
    val waitlistCount: Int = 0,
    val closed: Boolean = false,
)

@Serializable
data class DashboardDto(
    val period: String,
    val from: String,
    val to: String,
    val revenueCents: Long = 0,
    val revenueTrendPercent: Int = 0,
    val appointmentCount: Int = 0,
    val noShowPercent: Double = 0.0,
    val averageTicketCents: Long = 0,
    val operatorOccupancy: List<OperatorOccupancyDto> = emptyList(),
    val upcomingDays: List<UpcomingDayDto> = emptyList(),
    val inactiveClients: List<ClientDto> = emptyList(),
)

@Serializable
data class SalonNotificationSettingsDto(
    val bookingConfirmation: Boolean = true,
    val cancellationAlert: Boolean = true,
    val lateOperatorAlert: Boolean = false,
    val emptyDayPromos: Boolean = false,
)

@Serializable
data class ReminderRuleDto(val id: String, val hoursBefore: Int)

@Serializable
data class ReminderRulesEnvelope(val rules: List<ReminderRuleDto> = emptyList())

@Serializable
data class CreateReminderRuleBody(val hoursBefore: Int)

@Serializable
data class CampaignDto(
    val id: String,
    val name: String,
    val segment: String,
    val title: String,
    val body: String,
    val scheduledAt: String? = null,
    /** Orario da muro scelto dal titolare: è quello che la schermata rimostra. */
    val scheduledDate: String? = null,
    val scheduledTime: String? = null,
    val repeatWeekly: Boolean = false,
    val sendCap: Int? = null,
    val reachableCount: Int = 0,
    val segmentSize: Int = 0,
    val sentCount: Int = 0,
    val status: String = "DRAFT",
    val lastSentAt: String? = null,
)

@Serializable
data class CampaignsEnvelope(val campaigns: List<CampaignDto> = emptyList())

/**
 * Il corpo di una campagna.
 *
 * `scheduledAt` e `sendCap` sono [JsonElement] e non `String?`/`Int?` di
 * proposito: sul `PATCH` il backend distingue `null` ("togli la
 * programmazione") da campo assente ("lascia com'era"), e con
 * `explicitNulls = false` un `null` Kotlin sparirebbe dal JSON. Passandoci
 * `JsonNull` il `null` arriva davvero. Si costruiscono con `jsonOrNull(...)`.
 */
@Serializable
data class CampaignBody(
    val name: String,
    val segment: String,
    val title: String,
    val body: String,
    val scheduledAt: JsonElement,
    val repeatWeekly: Boolean = false,
    val sendCap: JsonElement,
    val status: String? = null,
)

/** Destinatari davvero raggiungibili di un segmento: li conta il server. */
@Serializable
data class CampaignReachDto(val reachable: Int = 0, val segmentSize: Int = 0)

File: ios/MenCare/Core/Network/Dto/AdminDto.swift
import Foundation

// Area titolare — docs/API.md §Gestione (solo titolare).

struct DashboardDto: Decodable {
    struct Occupancy: Decodable {
        let operatorId: String
        let operatorName: String
        let percent: Int
    }

    struct UpcomingDayDto: Decodable {
        let date: LocalDate
        let occupancyPercent: Int
        let waitlistCount: Int
        let closed: Bool
    }

    let period: String
    let from: LocalDate
    let to: LocalDate
    let revenueCents: Int64
    let revenueTrendPercent: Int
    let appointmentCount: Int
    let noShowPercent: Double
    let averageTicketCents: Int64
    let operatorOccupancy: [Occupancy]
    let upcomingDays: [UpcomingDayDto]
    let inactiveClients: [ClientDto]

    func toDomain() -> DashboardStats {
        DashboardStats(
            period: WireEnum.dashboardPeriod(period),
            revenueCents: revenueCents,
            revenueTrendPercent: revenueTrendPercent,
            appointmentCount: appointmentCount,
            noShowPercent: noShowPercent,
            averageTicketCents: averageTicketCents,
            operatorOccupancy: operatorOccupancy.map {
                OperatorOccupancy(operatorId: $0.operatorId, operatorName: $0.operatorName, percent: $0.percent)
            },
            upcomingDays: upcomingDays.map {
                UpcomingDay(date: $0.date, occupancyPercent: $0.occupancyPercent, waitlistCount: $0.waitlistCount, closed: $0.closed)
            },
            inactiveClients: inactiveClients.map { $0.toDomain() }
        )
    }
}

/// I quattro interruttori del salone; i promemoria stanno su una rotta a parte.
struct SalonNotificationSettingsDto: Codable {
    let bookingConfirmation: Bool
    let cancellationAlert: Bool
    let lateOperatorAlert: Bool
    let emptyDayPromos: Bool
}

struct ReminderRuleDto: Decodable {
    let id: String
    let hoursBefore: Int

    func toDomain() -> ReminderRule {
        ReminderRule(id: id, hoursBefore: hoursBefore)
    }
}

struct ReminderRulesResponseDto: Decodable {
    let rules: [ReminderRuleDto]
}

struct ReminderRuleWriteDto: Encodable {
    let hoursBefore: Int
}

struct CampaignDto: Decodable {
    let id: String
    let name: String
    let segment: String
    let title: String
    let body: String
    let scheduledAt: String?
    let scheduledDate: LocalDate?
    let scheduledTime: LocalTime?
    let repeatWeekly: Bool
    let sendCap: Int?
    let reachableCount: Int
    let segmentSize: Int
    let sentCount: Int
    let status: String
    let lastSentAt: String?

    func toDomain() -> PushCampaign {
        PushCampaign(
            id: id,
            name: name,
            segment: WireEnum.campaignSegment(segment),
            title: title,
            body: body,
            // Il server manda sia l'istante sia l'ora da muro: al titolare
            // interessa quella che ha scelto lui sul calendario.
            scheduledAt: scheduledDate.flatMap { date in scheduledTime.map { date.atTime($0) } },
            repeatWeekly: repeatWeekly,
            sendCap: sendCap,
            reachableCount: reachableCount,
            segmentSize: segmentSize,
            status: WireEnum.campaignStatus(status)
        )
    }
}

struct CampaignsResponseDto: Decodable {
    let campaigns: [CampaignDto]
}

struct CampaignWriteDto: Encodable {
    let name: String
    let segment: String
    let title: String
    let body: String
    /// Orario da muro del salone, `2026-10-02T09:00`: il server lo porta a
    /// istante con le stesse regole dell'agenda.
    let scheduledAt: String?
    let repeatWeekly: Bool
    let sendCap: Int?
    let status: String
}

struct CampaignReachDto: Decodable {
    let reachable: Int
    let segmentSize: Int
}

struct CampaignSendResultDto: Decodable {
    let queued: Int?
}
